subsys/net/lib/lwm2m/lwm2m_util.c

```diff
@@ -422,3 +423,132 @@ int lwm2m_ftoa(double *input, char *out, size_t outlen, int8_t dec_limit)
 			/* handle negative val2 when val1 is 0 */
 			(val1 == 0 && val2 < 0) ? "-" : "", (long long)val1, buf);
 }
+
+int lwm2m_path_to_string(char *buff, size_t buff_size, struct lwm2m_obj_path *input, int level_max)
+{
+	size_t fpl = 0; /* Length of the formed path */
+	int level;
+	int w;
+
+	if (!buff || buff_size < sizeof("/") || !input) {
+		return -EINVAL;
+	}
+
+	memset(buff, '\0', buff_size);
+
+	level = MIN(input->level, level_max);
+
+	/* Write path element at a time and leave space for the terminating NULL */
+	for (int idx = LWM2M_PATH_LEVEL_NONE; idx <= level; idx++) {
+		switch (idx) {
+		case LWM2M_PATH_LEVEL_NONE:
+			w = snprintk(&(buff[fpl]), buff_size - fpl, "/");
+			break;
+		case LWM2M_PATH_LEVEL_OBJECT:
+			w = snprintk(&(buff[fpl]), buff_size - fpl, "%" PRIu16 "/", input->obj_id);
+			break;
+		case LWM2M_PATH_LEVEL_OBJECT_INST:
+			w = snprintk(&(buff[fpl]), buff_size - fpl, "%" PRIu16 "/",
+				     input->obj_inst_id);
+			break;
+		case LWM2M_PATH_LEVEL_RESOURCE:
+			w = snprintk(&(buff[fpl]), buff_size - fpl, "%" PRIu16 "", input->res_id);
+			break;
+		case LWM2M_PATH_LEVEL_RESOURCE_INST:
+			w = snprintk(&(buff[fpl]), buff_size - fpl, "/%" PRIu16 "",
+				     input->res_inst_id);
+			break;
+		default:
+			__ASSERT_NO_MSG(false);
+			return -EINVAL;
+		}
+
+		if (w < 0) {
+			return -ENOBUFS;
+		}
+
+		/* Next path element, overwrites terminating NULL */
+		fpl += w;
+	}
+
+	return fpl;
+}
+
+uint16_t lwm2m_atou16(const uint8_t *buf, uint16_t buflen, uint16_t *len)
+{
+	uint16_t val = 0U;
+	uint16_t pos = 0U;
+
+	/* we should get a value first - consume all numbers */
+	while (pos < buflen && isdigit(buf[pos])) {
+		val = val * 10U + (buf[pos] - '0');
+		pos++;
+	}
+
+	*len = pos;
+	return val;
+}
+
+int lwm2m_string_to_path(const char *pathstr, struct lwm2m_obj_path *path,
+			  char delim)
+{
+	uint16_t value, len;
+	int i, tokstart = -1, toklen;
+	int end_index = strlen(pathstr) - 1;
+
+	(void)memset(path, 0, sizeof(*path));
+	for (i = 0; i <= end_index; i++) {
+		/* search for first numeric */
+		if (tokstart == -1) {
+			if (!isdigit((unsigned char)pathstr[i])) {
+				continue;
+			}
+
+			tokstart = i;
+		}
+
+		/* find delimiter char or end of string */
+		if (pathstr[i] == delim || i == end_index) {
+			toklen = i - tokstart + 1;
+
+			/* don't process delimiter char */
+			if (pathstr[i] == delim) {
+				toklen--;
+			}
+
+			if (toklen <= 0) {
+				continue;
+			}
+
+			value = lwm2m_atou16(&pathstr[tokstart], toklen, &len);
+			switch (path->level) {
+
+			case 0:
```

Comment:
Since we're already cleaning up, I think it'd be more readable to increment the path before the switch statement, and then use the corresponding path symbols, `LWM2M_PATH_LEVEL_OBJECT` etc.

Reply:
ah. Sorry I loose your comment here.